### PR TITLE
Auto: implement issue #627

### LINKED ISSUE #627: Add LLMService with multi-provider support
Subtask of #41
Depends on #626

Create src/services/llm_service.py with a unified LLM interface: chat(messages, model?) -> str, embed(text) -> list[float], configurable provider selection, retry logic with backoff, and cost tracking per tenant. Raises ValidationException on provider errors. Does NOT integrate into FastAPI deps yet. Add unit tests mocking the underlying provider API calls. Touches only src/services/llm_service.py and its test file.

Dev-Plan: docs/dev-plan/99-misc/0627-add-llmservice-with-multi-provider-support.md

### PR DESCRIPTION
**Pull request by yihua-zhuo**
Closes #627

Plan: `.plans/issue-627.md`

This PR was auto-implemented by Claude Code in three stages:

1. **Plan** — Claude wrote a detailed implementation plan to `.plans/issue-627.md`.
2. **Implement** — Claude executed the plan, ran `ruff check src/` + the unit and integration test suites, and iterated until all three were green.
3. **Self-review** — Claude reviewed its own diff against `codereview.md` (the project review checklist) and fixed any rule violations found, re-running the test suite if any changes were made.

PR-time CI (`Dev Agent CI`) re-verifies independently before this PR is mergeable.

Please review before merging.

## Release Notes

* **New Features**
  * Added unified LLM service supporting chat and embeddings across OpenAI and Anthropic providers
  * Automatic per-tenant cost tracking for LLM usage
  * Built-in retry logic with exponential backoff for improved reliability

* **Tests**
  * Comprehensive test coverage added for LLM service functionality

### CODE REVIEW

File: .plans/issue-627.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+## Source Contract
+Dev-plan target: `/home/runner/work/agent-job/agent-job/docs/dev-plan/99-misc/0627-add-llmservice-with-multi-provider-support.md`
+Template depth: `deep`
+Reading order followed:
+1. `/home/runner/work/agent-job/agent-job/docs/dev-plan/README.md`
+2. `/home/runner/work/agent-job/agent-job/docs/dev-plan/_template-deep.md`
+3. `/home/runner/work/agent-job/agent-job/docs/dev-plan/99-misc/0627-add-llmservice-with-multi-provider-support.md`
+
+## Affected Files
+- `src/services/llm_service.py` — new file; `LLMService` class with `__init__(session)`, `chat`, `embed`, `get_cost`, private `_call_openai` / `_call_anthropic`, module-level `OPENAI_API_URL` / `ANTHROPIC_API_URL` / `DEFAULT_MODEL` constants
```
Comment (yihua-zhuo) [inline .plans/issue-627.md:15]: **WARNING** (suggestion)

**Issue:** Plan lists only OPENAI_API_URL/ANTHROPIC_API_URL/DEFAULT_MODEL as module constants, but the implementation also defines MAX_RETRIES, OPENAI_COST_PER_M_TOKEN, ANTHROPIC_COST_PER_M_TOKEN, EMBEDDING_COST_PER_M_TOKEN — the plan should enumerate every module-level constant it introduces (Rule 140).

**Suggestion:** Update the Affected Files / Implementation Steps to list all seven module constants.

File: .plans/issue-627.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+- `tests/unit/test_llm_service.py` — new file; unit tests mocking `httpx.AsyncClient.post` via `unittest.mock.AsyncMock`
+
+## Implementation Steps
+
+1. **Create `src/services/llm_service.py` with the class skeleton and module constants.**
+   - Imports: `os`, `asyncio`, `httpx`, `from sqlalchemy.ext.asyncio import AsyncSession`, `from pkg.errors.app_exceptions import ValidationException`.
+   - Module constants: `OPENAI_API_URL = "https://api.openai.com/v1/chat/completions"`, `ANTHROPIC_API_URL = "https://api.anthropic.com/v1/messages"`, `DEFAULT_MODEL = "gpt-4o"`, `MAX_RETRIES = 3`.
```
Comment (yihua-zhuo) [inline .plans/issue-627.md:22]: **WARNING** (suggestion)

**Issue:** Plan requires `__init__(self, session: AsyncSession)` for a service that never uses the session, embedding a dead dependency into the interface — the plan should either justify the session parameter or drop it (Rule 140, Rule 17).

**Suggestion:** Remove the `session` parameter from `__init__` since the service makes no DB calls; cost tracking is in-memory only.

File: .plans/issue-627.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   - `class LLMService: __init__(self, session: AsyncSession)` stores `self.session`, instantiates `self._client = httpx.AsyncClient(timeout=30.0)`, and initializes `self._cost_by_tenant: dict[int, float] = {}`.
+
+2. **Implement `_call_openai(self, payload: dict, tenant_id: int) -> dict` and `_call_anthropic(self, payload: dict, tenant_id: int) -> dict`.**
+   - Each runs a `for attempt in range(MAX_RETRIES)` loop calling `self._client.post(url, json=payload, headers=...)` (OpenAI: `Authorization: Bearer OPENAI_API_KEY`; Anthropic: `x-api-key` + `anthropic-version` header).
+   - On `status_code == 200`, parse JSON; pull `usage.total_tokens` from OpenAI / `usage.input_tokens + output_tokens` from Anthropic; increment `self._cost_by_tenant[tenant_id]` by `(tokens / 1_000_000) * cost_per_m_token` (use a hardcoded per-model rate constant, e.g. `OPENAI_COST_PER_M_TOKEN = 0.002`); return the parsed dict.
+   - On non-200, `await asyncio.sleep(2 ** attempt)` if `attempt < MAX_RETRIES - 1`; after the loop exits without success, raise `ValidationException(f"LLM provider error: {provider_name} returned {status_code} after {MAX_RETRIES} retries")`.
```
Comment (yihua-zhuo) [inline .plans/issue-627.md:28]: **WARNING** (suggestion)

**Issue:** Plan says the retry loop fires on any non-200 status, including 4xx client errors (e.g. 400, 401) that will never succeed on retry — this is an unbounded-waste pattern the plan should call out and restrict (Rule 42).

**Suggestion:** Limit retries to transient 5xx and 429 status codes; do not retry 4xx client errors.

File: .plans/issue-627.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   - On success, return `resp.json()["data"][0]["embedding"]`; on exhaustion, raise `ValidationException(f"LLM provider error: OpenAI embeddings returned {status_code} after {MAX_RETRIES} retries")`.
+   - Update `self._cost_by_tenant[tenant_id]` by the token count from `usage.total_tokens` using the OpenAI embedding rate.
+
+5. **Implement `get_cost(self, tenant_id: int) -> float` returning `self._cost_by_tenant.get(tenant_id, 0.0)`.**
+
+6. **Add `__aenter__` and `__aexit__` to the class** so the service can be used as `async with LLMService(session) as svc:` — `__aenter__` returns `self`, `__aexit__` calls `await self._client.aclose()`. This prevents the `httpx.AsyncClient` leak noted in the dev-plan §4.4 known-pitfall #2.
+
+7. **Create `tests/unit/test_llm_service.py` with 8 unit tests covering:**
+   - `test_chat_openai_happy_path` — mock `httpx.AsyncClient.post` to return 200 + OpenAI-shaped JSON (`{"choices": [{"message": {"content": "hi"}}], "usage": {"total_tokens": 100}}`); assert `chat()` returns `"hi"` and `get_cost(tenant_id)` reflects the token charge.
```
Comment (yihua-zhuo) [inline .plans/issue-627.md:45]: **INFO** (suggestion)

**Issue:** Plan lists 8 unit tests, and the file has 8 test functions — counts match — but no test covers the context-manager lifecycle (closing the httpx client on __aexit__), a missing API key, or a malformed usage payload (Rule 64, Rule 45).

**Suggestion:** Add tests for: client closed on __aexit__, empty/missing API key env var, and response missing the 'usage' key.

File: src/services/llm_service.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+OPENAI_COST_PER_M_TOKEN = 0.002
+ANTHROPIC_COST_PER_M_TOKEN = 0.003
+EMBEDDING_COST_PER_M_TOKEN = 0.00002
+
+
+class LLMService:
+    """Unified multi-provider LLM access — chat, embed, and per-tenant cost tracking."""
+
+    def __init__(self, session: AsyncSession):
+        self.session = session
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:25]: **WARNING** (suggestion)

**Issue:** `self.session` is stored but never read — dead reference, and the service is forced to take an `AsyncSession` it does not use, violating Rule 17 (dead code) and Rule 14 (misleading abstraction).

**Suggestion:** Drop the `session` parameter from `__init__` (and update tests), or use the session for cost persistence if that is the intent.

File: src/services/llm_service.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    ) -> list[float]:
+        """Return embedding vector for the given text. Raise ValidationException on provider error."""
+        payload = {"model": model, "input": text}
+        url = f"{OPENAI_API_URL.rsplit('/', 1)[0]}/embeddings"
+        headers = {
+            "Authorization": f"Bearer {os.environ.get('OPENAI_API_KEY', '')}",
+            "Content-Type": "application/json",
+        }
+
+        for attempt in range(MAX_RETRIES):
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:67]: **WARNING** (suggestion)

**Issue:** Retry loop fires on every non-200 status, including 4xx client errors (400/401/403) that will never recover on retry — wastes time and budget, violates Rule 42 (retry logic is bounded and safe).

**Suggestion:** Restrict the retry condition: `if 500 <= resp.status_code or resp.status_code == 429` (or similar); let 4xx raise immediately.

File: src/services/llm_service.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                await asyncio.sleep(2**attempt)
+        raise ValidationException(
+            f"LLM provider error: OpenAI embeddings returned {resp.status_code} after {MAX_RETRIES} retries"
+        )
+
+    async def get_cost(self, tenant_id: int) -> float:
+        """Return accumulated LLM cost for this tenant in USD."""
+        return self._cost_by_tenant.get(tenant_id, 0.0)
+
+    async def _call_openai(self, payload: dict, tenant_id: int) -> dict:
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:86]: **WARNING** (suggestion)

**Issue:** `_call_openai`, `_call_anthropic`, and the `embed` method each duplicate the identical retry-loop / cost-update / raise-on-exhaustion body — three near-copies of ~15 lines each, violating Rule 110.

**Suggestion:** Extract a private `_post_with_retry(url, payload, headers, tenant_id, cost_per_m_token) -> dict` helper and call it from all three sites.

File: src/services/llm_service.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+ANTHROPIC_COST_PER_M_TOKEN = 0.003
+EMBEDDING_COST_PER_M_TOKEN = 0.00002
+
+
+class LLMService:
+    """Unified multi-provider LLM access — chat, embed, and per-tenant cost tracking."""
+
+    def __init__(self, session: AsyncSession):
+        self.session = session
+        self._client = httpx.AsyncClient(timeout=30.0)
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:26]: **WARNING** (suggestion)

**Issue:** An `httpx.AsyncClient(timeout=30.0)` is opened in `__init__` with no explicit `close()` method — callers who instantiate `LLMService(session)` without `async with` will leak the client and its connection pool, violating Rule 115 (resource lifecycle is explicit).

**Suggestion:** Add a public `async def close()` that calls `await self._client.aclose()`, and document that both context-manager use and manual `close()` are valid.

Comment (yihua-zhuo) [inline src/services/llm_service.py:26]: **INFO** (suggestion)

**Issue:** Per-request timeout of 30 s combined with up to 2 backoff sleeps of 1 s + 2 s means a failed request can occupy a request handler for ~93 s — no overall deadline on the retry loop (Rule 44, Rule 114).

**Suggestion:** Track elapsed time in the retry loop and bail out when a configurable overall deadline (e.g. 15 s) is exceeded.

File: src/services/llm_service.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    (tokens / 1_000_000) * EMBEDDING_COST_PER_M_TOKEN
+                )
+                return data["data"][0]["embedding"]
+            if attempt < MAX_RETRIES - 1:
+                await asyncio.sleep(2**attempt)
+        raise ValidationException(
+            f"LLM provider error: OpenAI embeddings returned {resp.status_code} after {MAX_RETRIES} retries"
+        )
+
+    async def get_cost(self, tenant_id: int) -> float:
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:82]: **INFO** (suggestion)

**Issue:** `get_cost` is declared `async` but performs only a dict lookup — no await, no IO; violates Rule 29 (correct async/await usage) and forces callers to `await` a synchronous read.

**Suggestion:** Make `get_cost` a plain `def`.

File: src/services/llm_service.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from pkg.errors.app_exceptions import ValidationException
+
+OPENAI_API_URL = "https://api.openai.com/v1/chat/completions"
+ANTHROPIC_API_URL = "https://api.anthropic.com/v1/messages"
+DEFAULT_MODEL = "gpt-4o"
+MAX_RETRIES = 3
+
+OPENAI_COST_PER_M_TOKEN = 0.002
```
Comment (yihua-zhuo) [inline src/services/llm_service.py:16]: **INFO** (suggestion)

**Issue:** The three per-token cost constants (`OPENAI_COST_PER_M_TOKEN`, `ANTHROPIC_COST_PER_M_TOKEN`, `EMBEDDING_COST_PER_M_TOKEN`) are hardcoded module globals — the plan acknowledges this is a stub, but they will silently rot as provider pricing changes (Rule 19, Rule 109).

**Suggestion:** Source pricing from settings/config with a documented override mechanism, and add a TODO pointing to a pricing file.

File: tests/unit/test_llm_service.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+
+async def test_chat_openai_happy_path(session):
+    ok_body = {
+        "choices": [{"message": {"content": "hi"}}],
+        "usage": {"total_tokens": 100},
+    }
+    with patch("services.llm_service.httpx.AsyncClient.post", new_callable=AsyncMock) as mock_post:
+        mock_post.return_value = _make_response(200, ok_body)
+        async with LLMService(session) as svc:
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:42]: **WARNING** (suggestion)

**Issue:** Test holds a reference to `svc` after the `async with` block has already called `__aexit__` and closed the httpx client — the cost assertion works only because `get_cost` is dict-only, but the pattern is misleading; it also means no test actually exercises post-close behavior or the cleanup path itself (Rule 67, Rule 46).

**Suggestion:** Move the cost assertion inside the `async with` block, and add a dedicated test that calls `chat()` after `__aexit__` to assert the client is truly closed.

File: tests/unit/test_llm_service.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+def _make_response(status_code: int, json_data: dict | None = None) -> MagicMock:
+    """Build a mock httpx response with given status and JSON body."""
+    resp = MagicMock()
+    resp.status_code = status_code
+    resp.json = MagicMock(return_value=json_data or {})
+    return resp
+
+
+@pytest.fixture
+def session():
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:26]: **INFO** (suggestion)

**Issue:** The `session` fixture is an `AsyncMock()` passed to a service that never uses the session — passing it in the first place validates the dead-parameter contract rather than the service's actual surface (Rule 66).

**Suggestion:** Once the `session` parameter is removed, drop the fixture and instantiate `LLMService()` directly.

File: tests/unit/test_llm_service.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    return AsyncMock()
+
+
+# ---------------------------------------------------------------------------
+# Happy path: OpenAI chat
+# ---------------------------------------------------------------------------
+
+
+async def test_chat_openai_happy_path(session):
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:35]: **INFO** (suggestion)

**Issue:** No test covers a missing/empty API key env var (the code silently sends `Bearer ` with an empty token, which will produce a 401 and 3 useless retries) — Rule 45 (failure paths are tested) and Rule 64 (edge cases).

**Suggestion:** Add a test that unsets `OPENAI_API_KEY` / `ANTHROPIC_API_KEY` and asserts a clear error is raised, or assert the headers carry the expected (possibly empty) token.

File: tests/unit/test_llm_service.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    expected_cost = (100 / 1_000_000) * ANTHROPIC_COST_PER_M_TOKEN
+    assert await svc.get_cost(1) == expected_cost
+
+
+# ---------------------------------------------------------------------------
+# Embedding
+# ---------------------------------------------------------------------------
+
+
+async def test_embed_returns_vector(session):
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:77]: **INFO** (suggestion)

**Issue:** No test covers a response with a missing `usage` key — the code uses `.get(..., 0)` so the call succeeds silently with zero cost; this should be asserted as intentional (Rule 45).

**Suggestion:** Add a test that returns a 200 response with no `usage` block and asserts `get_cost` returns 0.0 and the call still succeeds.

File: tests/unit/test_llm_service.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    expected_cost = (10 / 1_000_000) * EMBEDDING_COST_PER_M_TOKEN
+    assert await svc.get_cost(1) == expected_cost
+
+
+# ---------------------------------------------------------------------------
+# Retry success
+# ---------------------------------------------------------------------------
+
+
+async def test_chat_retry_on_429(session):
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:99]: **INFO** (suggestion)

**Issue:** No test asserts that 4xx client errors (e.g. 400, 401) are NOT retried — the production code retries all non-200, so a test that documents the expected behavior would lock the contract in (Rule 45, Rule 64).

**Suggestion:** Add a test that returns 400 and asserts either immediate failure (if you fix the code) or `call_count == 3` (if you document the current behavior) — the current absence of a test means the behavior is unspecified.

File: tests/unit/test_llm_service.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+# Happy path: OpenAI chat
+# ---------------------------------------------------------------------------
+
+
+async def test_chat_openai_happy_path(session):
+    ok_body = {
+        "choices": [{"message": {"content": "hi"}}],
+        "usage": {"total_tokens": 100},
+    }
+    with patch("services.llm_service.httpx.AsyncClient.post", new_callable=AsyncMock) as mock_post:
```
Comment (yihua-zhuo) [inline tests/unit/test_llm_service.py:40]: **INFO** (suggestion)

**Issue:** Patches `httpx.AsyncClient.post` at the class level — works, but a more precise patch on the instance's `post` would avoid leaking the mock to any other `AsyncClient` created in the same test process (Rule 135).

**Suggestion:** Patch `LLMService.__init__` to inject the mock client, or assign `svc._client.post = AsyncMock(...)` and skip the class-level patch.

**Comment by yihua-zhuo** [comment]
## Claude Code Review (inline)

**Status:** `pass`
**Inline comments posted:** 17

**Summary:** Reviewed 3 file(s) between origin/master and origin/auto/issue-627.
